src/diffusers/models/attention_dispatch.py

```diff
@@ -0,0 +1,917 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import contextlib
+import inspect
+from enum import Enum
+from typing import Any, Callable, Dict, List, Literal, Optional, Tuple, Union
+
+import torch
+
+from ..utils import (
+    OptionalDependencyNotAvailable,
+    get_logger,
+    is_flash_attn_available,
+    is_flash_attn_version,
+    is_sageattention_available,
+    is_sageattention_version,
+    is_torch_version,
+    is_xformers_available,
+    is_xformers_version,
+)
+from ..utils.constants import DIFFUSERS_ATTN_CHECKS, DIFFUSERS_ATTN_PROVIDER
+
+
+if is_flash_attn_available():
+    if is_flash_attn_version("<", "2.6.3"):
+        raise OptionalDependencyNotAvailable(
+            "The `flash-attn` library version is too old. Please update it to at least 2.6.3."
+        )
+
+    from flash_attn import flash_attn_func, flash_attn_varlen_func
+else:
+    flash_attn_func = None
+    flash_attn_varlen_func = None
+
+
+if is_sageattention_available():
+    if is_sageattention_version("<", "2.1.1"):
+        raise OptionalDependencyNotAvailable(
+            "The `sageattention` library version is too old. Please update it to at least 2.1.1."
+        )
+
+    from sageattention import (
+        sageattn,
+        sageattn_qk_int8_pv_fp8_cuda,
+        sageattn_qk_int8_pv_fp8_cuda_sm90,
+        sageattn_qk_int8_pv_fp16_cuda,
+        sageattn_qk_int8_pv_fp16_triton,
+        sageattn_varlen,
+    )
+else:
+    sageattn = None
+    sageattn_qk_int8_pv_fp16_cuda = None
+    sageattn_qk_int8_pv_fp16_triton = None
+    sageattn_qk_int8_pv_fp8_cuda = None
+    sageattn_qk_int8_pv_fp8_cuda_sm90 = None
+    sageattn_varlen = None
+
+
+if is_torch_version(">=", "2.5.0"):
+    from torch.nn.attention.flex_attention import BlockMask, create_block_mask
+    from torch.nn.attention.flex_attention import flex_attention as torch_flex_attention
+else:
+    create_block_mask = None
+    torch_flex_attention = None
+
+    class BlockMask:
+        def __init__(self, *args, **kwargs):
+            raise OptionalDependencyNotAvailable(
+                "The `torch` library version is too old. Please update it to at least 2.5.0."
```

Comment:
We could further clarify that "To use `BlockMask` you need an updated torch installation."